infer/src/pulse/PulseExecutionDomain.ml

```diff
@@ -102,6 +108,170 @@ let pp fmt exec_state =
   F.fprintf fmt "%a%a" (pp_header TEXT) exec_state AbductiveDomain.pp (to_astate exec_state)
 
 
+(* Infinite loop detector creates a new InfiniteProgram state when detected *)
+
+let widenstate = AnalysisGlobalState.make_dls ~init:(fun () -> Some (Stdlib.Hashtbl.create 16))
+
+let get_widenstate () = DLS.get widenstate
+
+let add_widenstate key =
+  Utils.with_dls widenstate ~f:(fun widenstate ->
+      match widenstate with
+      | None ->
+          widenstate
+      | Some state ->
+          Stdlib.Hashtbl.add state key () ;
+          widenstate )
+
+
+let rec find_infinite_state (lst : t list) : bool =
+  match lst with
+  | hd :: tl -> (
+    match hd with InfiniteProgram _ -> true | _ -> find_infinite_state tl )
+  | _ ->
+      false
+
+
+let back_edge (prev : t list) (next : t list) (num_iters : int) : t list * int =
+  let has_infinite_state = find_infinite_state next in
+  if has_infinite_state then ([], -1)
+  else
+    let _ = num_iters in
+    let cfgnode = AnalysisState.get_node_exn () in
+    let same = phys_equal prev next in
+    let rec detect_elem e lst curi : bool * int =
+      match lst with
+      | [] ->
+          (false, -1)
+      | hd :: tl ->
+          if leq ~lhs:hd ~rhs:e then (true, curi) else detect_elem e tl (curi + 1)
+    in
+    let rec compute_workset prev next newset nextit : (t * int) list =
+      match next with
+      | [] ->
+          newset
+      | hd :: tl ->
+          let res, idx = detect_elem hd prev 0 in
+          let recidx = match idx with -1 -> nextit | _ -> idx in
+          if res then compute_workset prev tl newset (nextit + 1)
+          else [(hd, recidx)] @ compute_workset prev tl newset (nextit + 1)
```

Comment:
I'm confused why the index `recidx` sometimes corresponds to the index in the `prev` list and sometimes in the `next` list.

I think `detect_elem` can be replaced by `List.findi`. And maybe `cmpute_workset` itself can be replaced by a `List.filter_mapi`?
```suggestion
          match List.findi prev ~f:(fun _i prev_disjunct ->leq ~lhs:prev_disjunct ~rhs:hd) with
          | None -> (hd, nextit)::compute_workset prev tl newset (nextit + 1)
          | Some (_, idx) ->compute_workset prev tl newset (nextit + 1)
```